### ISSUE TITLE
fixup delete tracking to allow re-insertion of the deleted item

### ISSUE THREAD

**Issue by tantaman** [vlcn-io/cr-sqlite#71]
See discussion in #59

**Comment by tantaman** [comment]
You can work around this in user space by keeping an "is_deleted" column and toggling that rather than actually deleting the row

**Comment by tantaman** [comment]
We can also fix this by adding create records and recording the version at create time.

**Comment by Azarattum** [comment]
The diagram of an implementation we've discussed on Discord:
![image](https://user-images.githubusercontent.com/43073346/213863893-8eb8ca61-12e3-4f73-979d-06ac748bae03.png)

**Comment by tantaman** [comment]
Some other things to consider as me move to implement delete tracking --

> You can work around this in user space by keeping an "is_deleted" column and toggling that rather than actually deleting the row

To explain that a bit more, tables would have boolean `is_deleted` columns

```sql
CREATE TABLE todo (id primary key, content text, complete boolean, is_deleted boolean);
```

Rather than issuing deletes, the app would toggle `is_deleted` to `true` or `false` depending on the delete state (given this is a boolean value, the `is_deleted` version column effectively is a [causal length](https://www.researchgate.net/publication/340954399_A_low-cost_set_CRDT_based_on_causal_lengths)). User space is of course now required to filter against `is_deleted` when retrieving rows.

User space / soft-delete like this actually has a number of benefits for the application developer.

1. We can preserve referential integrity
2. Applications can surface delete conflicts

## 1 - Preserve referential integrity

Given the rows will still exist post-delete, foreign key constraints can be re-introduced. A merge will never break a foreign key constraint. 

Note to self: we'll need to update create syncing to sync _the entire row_ any time a create event is hit.

## 2 - Surface delete conflicts

This is a case I've been hitting frequently in real world apps. There are times where a user is working on something that another user deletes. If you _actually_ delete the thing, the user that was working on it gets a rather jarring experience and the application is often left in an inconsistent state.

Examples:
- In a presentation editor the slide I'm working on would just disappear and I'm left with no slide selected or my "selected state" pointing to a slide that doesn't exist and the app logic having to handle this. All links to the now deleted slide are also broken if slides can link to one another. The app now needs to handle, in every place, the possibility of a retrieved row not existing.
- In a music app, my playlist could disappear leading to the same dangling reference issues and weird experience

`is_deleted` prevents any dangling references and allows the app to provide a UI affordance to the user letting them know that the current thing is deleted and they can un-delete it if desired.


If crsqlite hides all the mechanics of deletion then app developers lose any abilities to create UI affordances around delete.

**Comment by tantaman** [comment]
Discussed deletes with Paul and @ivertom today.

Paul had an interesting approach to consistently handling deletes while also preserving referential integrity.

The basic idea was to cascade the delete. When replicating at another node, cascade the delete there too in order to clear out any new references that were created.

If a new reference pops up from a different peer and is merged to you and the reference check fails, don't take their insert. The other peer will eventually lose their insert too when they receive the cascade delete update.

**Comment by tantaman** [comment]
Options for delete:

1. Soft delete (preserves referential integrity, allows ui affordance, simple)
2. Cascade delete (preserves referential integrity, hard?, makes undo support really hard?)
3. Delete wins, no cascade (breaks ref integrity, simple)
4. Delete + undo on FK constraint failure (Iver's thesis takes this approach. Preserves referential integrity but might get into update loops?)
5. Add wins + adding FK touches thing pointed to (called compensations in the literature)
6. Add wins w/ no FK compensations
7. Other?

tldr -- delete has a surprising number of ways it can be handled and what you want is likely to differ by use case.

**Comment by Azarattum** [comment]
I'm generally against soft deletes (at least by default). First of all, it is really bad for privacy purposes. If a user deletes something, they should be sure that the data is gone for good. `cr-sqlite` doesn't support undo for modification (since it's keeping only the last modification value). I don't see why deletes should be an exception. I actually see this as a strength, since most application don't need the undo functionality and the ones that do could implement it on an application level. I've used YJS where there is no way to opt out from keeping history afaik. A DB becomes bloated really fast there...

Cascade deletes would generally be my preferred option. I wonder if it would be possible to support [ON DELETE CASCADE](https://www.sqlite.org/foreignkeys.html) upon schema definition.

Current approach with not allowing foreign keys at all and leaving it to the userland is also fine. As long as the proposed method of reinserting would function properly, I can deal with time-to-time chores to go though entire database and strip all the hanging pointers.

I don't have any opinions on options 4-6. Maybe it would be easier to understand their usefulness with a real world use case example. I can't think of one on top of my head.

Generally it would be nice to support different strategies and allow a user to choose a preferred one. Maybe support the native SQLite [`ON DELETE` statement](https://www.sqlite.org/foreignkeys.html)? I'm not sure how hard that would be to implement though.

**Comment by tantaman** [comment]
Implementation sketch:

We will track whether or not a row exists by introducing a ["causal length"](https://dl.acm.org/doi/pdf/10.1145/3380787.3393678) metadata field for that row. This is basically an "is_deleted" flag in the metadata layer except that the flag is incremented by one for each delete/insert operation. I guess you can think of it as a compressed form of `is_deleted` + the `is_deleted` clock.

All row data is still removed on delete (with the exception of this flag and the primary keys used to identify the row). This new flag is used to understand whether or not a merge should resurrect the row. When merging, the largest causal length wins. An even causal length means cause the row to be deleted, an odd causal length causes it to be created. 

1. Change the sentinels (__crsql_pk_only and __crsql_del) to `__crsql_cl`
2. All insertions will create a `__crsql_cl` record for the row if it does not exist. The `cl` will default to 1.
3. Updates will not modify `__crsql_cl`
4. Given all insertions create a `__crsql_cl` we don't need special cased `pk_only` handling any more
5. Deletes will increment `__crsql_cl` if the row is not already deleted
6. Selecting from `crsql_changes` will return a column for `__crsql_cl` in-line with every row
7. Merging will resurrect or delete the row if the incoming causal length wins

Resurrection will have the same problematic? behavior of inserts in that they are not whole row insertions.

Whole row insertions would be a "next step", something in "v2", or something taken care of in networking layers.

**Comment by tantaman** [comment]
Started on this here: https://github.com/vlcn-io/cr-sqlite/tree/re-insert

Some updates to the merge algorithm are required.

We need to consider cases on merge where:
- `peer_row.causal_length % 2 == our_row.causal_length %2 && peer_row.causal_length > our_row.causal_length`

In this case, we missed a delete event. In this case, we should _delete all our data for the row_ then apply the update currently received for the row.

Another way to put it: comparing whether or not we take a value for a column is now a function of `causal_length` and `column_version`

```
if peer.causal_length > our.causal_length {
  if is_delete(peer.causal_length) {
    delete the row
  } else {
     take the peer's columns no-matter what our column versions are
  }
} else if peer.causal_length < our.causal_length {
  keep our columns (or lack of columns) no-matter what the peer's column versions are
} else {
  causal lengths are the same.
  revert to current algorithm for merging columns.
}
```

**Comment by tantaman** [comment]
Given this was touching so much code, and we eventually want to be writing in Rust, I ended up porting all the files that'd need to be changed to Rust.

This is now being worked on in https://github.com/vlcn-io/cr-sqlite/pull/292 and subsequent PRs stacked on that one.

**Comment by tantaman** [comment]
On to step 2 of 3 -- #295  collapse row metadata to a single entry. More background is in the description of pr #292.

**Comment by tantaman** [comment]
Ok, this is the last PR! https://github.com/vlcn-io/cr-sqlite/pull/299

Once I get existing tests to work and write some new ones we can mar this complete 🎉

The property based merge tests already work so that's a good sign 🥳

**Comment by tantaman** [comment]
> The property based merge tests already work so that's a good sign 🥳

Actually, after uncovering several bugs in this pr, I take this as a bad sign now and that these tests need to be more robust. Adding several new property based tests.
